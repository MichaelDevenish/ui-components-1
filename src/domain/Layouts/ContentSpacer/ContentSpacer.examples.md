#### Content Spacer

The content spacer is used to add a margin to the bottom of content items to space them out using standardised spacing.

The last item will always have no spacing (even if a spacing size is defined); this is to accomodate for situations where the last item may not be rendered (eg. because the user doesn't have permission to see it).

```jsx
const { Variables } = require('@Common');

<ContentSpacer
  contentItems={[
    {
      content: <div style={{backgroundColor: Variables.Color.n300, height: '200px'}}/>
    },
    {
      content: <div style={{backgroundColor: Variables.Color.n400, height: '200px'}}/>,
      spacingSize: 'small'
    },
    {
      content: <div style={{backgroundColor: Variables.Color.n500, height: '200px'}}/>,
      spacingSize: 'medium'
    },
    {
      content: <div style={{backgroundColor: Variables.Color.n600, height: '200px'}}/>,
      spacingSize: 'large'
    },
    {
      content: <div style={{backgroundColor: Variables.Color.n700, height: '200px'}}/>,
      spacingSize: 'large'
    }
  ]}
/>
```
