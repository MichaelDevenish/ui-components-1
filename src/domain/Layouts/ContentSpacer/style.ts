import styled from 'styled-components'

interface IContentSpacerContentItemProps {
  /** Size of the space between this content item and the next */
  spacingSize?: 'small' | 'medium' | 'large'
}

function spacingForContentItem(props: IContentSpacerContentItemProps): number {
  switch (props.spacingSize) {
    case 'small':
      return 24
    case 'medium':
      return 48
    case 'large':
      return 80
    default:
      return 0
  }
}

const StyledContentItem = styled.div<IContentSpacerContentItemProps>`
  margin-bottom: ${spacingForContentItem}px;

  :last-child {
    margin-bottom: 0px;
  }
`

export {
  StyledContentItem
}
