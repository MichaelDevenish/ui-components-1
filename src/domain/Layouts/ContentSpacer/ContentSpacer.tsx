import React from 'react'
import { StyledContentItem } from './style'
import { Props } from '../../../common/'

interface IContentSpacerContentItemProps {
  /** Size of the space between this content item and the next */
  size?: 'small' | 'medium' | 'large',
  /** Content item */
  content: JSX.Element,
  /** Component context for the content item */
  componentContext?: string
}

interface IContentSpacerProps {
  /** Provide an array of items for the content */
  contentItems: IContentSpacerContentItemProps[],
  /** Component context for the overall content spacer*/
  componentContext?: string
}

class ContentSpacer extends React.Component<IContentSpacerProps> {
  public render (): JSX.Element {
    const {
      contentItems,
      componentContext
    } = this.props

    return (
      <div
        data-component-type={Props.ComponentType.ContentSpacer}
        data-component-context={componentContext}
      >
        {contentItems.map((contentItem: IContentSpacerContentItemProps, index: number) => {
          return (
            <StyledContentItem
              size={contentItem.size}
              key={index}
              data-component-type={Props.ComponentType.ContentSpacerItem}
              data-component-context={contentItem.componentContext}
            >
              {contentItem.content}
            </StyledContentItem>
          )
        })}
      </div>
    )
  }
}

export {
  ContentSpacer,
  IContentSpacerContentItemProps
}
