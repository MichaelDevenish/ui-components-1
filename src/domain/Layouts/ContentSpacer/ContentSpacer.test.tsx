import { mount } from 'enzyme'
import React from 'react'

import { ContentSpacer } from './ContentSpacer'

describe('<ContentSpacer />', () => {
  it(`should render a content spacer with varying sizes`, () => {
    const wrapper = mount(
      <ContentSpacer
        contentItems={[
          {
            content: <div>a</div>,
            size: 'small'
          },
          {
            content: <div>b</div>,
            size: 'medium'
          },
          {
            content: <div>c</div>,
            size: 'large'
          },
          {
            content: <div>d</div>
          }
        ]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a content spacer with no sizes defined`, () => {
    const wrapper = mount(
      <ContentSpacer
        contentItems={[
          {
            content: <div>a</div>
          },
          {
            content: <div>b</div>
          },
          {
            content: <div>c</div>
          },
          {
            content: <div>d</div>
          }
        ]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
