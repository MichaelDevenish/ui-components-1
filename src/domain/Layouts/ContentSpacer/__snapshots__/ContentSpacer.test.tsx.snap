// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentSpacer /> should render a content spacer with no sizes defined 1`] = `
.c0 {
  margin-bottom: 0px;
}

<ContentSpacer
  contentItems={
    Array [
      Object {
        "content": <div>
          a
        </div>,
      },
      Object {
        "content": <div>
          b
        </div>,
      },
      Object {
        "content": <div>
          c
        </div>,
      },
      Object {
        "content": <div>
          d
        </div>,
      },
    ]
  }
>
  <div
    data-component-type="content_spacer"
  >
    <styled.div
      data-component-type="content_spacer_item"
      key="0"
    >
      <div
        className="c0"
        data-component-type="content_spacer_item"
      >
        <div>
          a
        </div>
      </div>
    </styled.div>
    <styled.div
      data-component-type="content_spacer_item"
      key="1"
    >
      <div
        className="c0"
        data-component-type="content_spacer_item"
      >
        <div>
          b
        </div>
      </div>
    </styled.div>
    <styled.div
      data-component-type="content_spacer_item"
      key="2"
    >
      <div
        className="c0"
        data-component-type="content_spacer_item"
      >
        <div>
          c
        </div>
      </div>
    </styled.div>
    <styled.div
      data-component-type="content_spacer_item"
      key="3"
    >
      <div
        className="c0"
        data-component-type="content_spacer_item"
      >
        <div>
          d
        </div>
      </div>
    </styled.div>
  </div>
</ContentSpacer>
`;

exports[`<ContentSpacer /> should render a content spacer with varying sizes 1`] = `
.c0 {
  margin-bottom: 24px;
}

.c1 {
  margin-bottom: 48px;
}

.c2 {
  margin-bottom: 80px;
}

.c3 {
  margin-bottom: 0px;
}

<ContentSpacer
  contentItems={
    Array [
      Object {
        "content": <div>
          a
        </div>,
        "size": "small",
      },
      Object {
        "content": <div>
          b
        </div>,
        "size": "medium",
      },
      Object {
        "content": <div>
          c
        </div>,
        "size": "large",
      },
      Object {
        "content": <div>
          d
        </div>,
      },
    ]
  }
>
  <div
    data-component-type="content_spacer"
  >
    <styled.div
      data-component-type="content_spacer_item"
      key="0"
      size="small"
    >
      <div
        className="c0"
        data-component-type="content_spacer_item"
        size="small"
      >
        <div>
          a
        </div>
      </div>
    </styled.div>
    <styled.div
      data-component-type="content_spacer_item"
      key="1"
      size="medium"
    >
      <div
        className="c1"
        data-component-type="content_spacer_item"
        size="medium"
      >
        <div>
          b
        </div>
      </div>
    </styled.div>
    <styled.div
      data-component-type="content_spacer_item"
      key="2"
      size="large"
    >
      <div
        className="c2"
        data-component-type="content_spacer_item"
        size="large"
      >
        <div>
          c
        </div>
      </div>
    </styled.div>
    <styled.div
      data-component-type="content_spacer_item"
      key="3"
    >
      <div
        className="c3"
        data-component-type="content_spacer_item"
      >
        <div>
          d
        </div>
      </div>
    </styled.div>
  </div>
</ContentSpacer>
`;
